src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -266,30 +288,54 @@ void CoreWorkerDirectActorTaskSubmitter::PushActorTask(const ClientQueue &queue,
 
   const auto task_id = task_spec.TaskId();
   const auto actor_id = task_spec.ActorId();
-  const auto counter = task_spec.ActorCounter();
+  const auto actor_counter = task_spec.ActorCounter();
+  const auto task_skipped = task_spec.GetMessage().skip_execution();
```

Comment:
What do you think about not having the task_skipped flag and letting the actor re-execute the out-of-order tasks? I could see use cases for both, so I'd prefer to start with no flag.

Reply:
The out of order tasks can be potentially very expensive to execute and mutate the state in undesirable way. Two examples in Serve:

## Serve Controller
1. `__init__` # reconstruct the state from internal kv upon restart
2. `start_control_loop()` # long running
3. `create_backend()` # state mutating and starting new actors, but we checkpoint the state to internal kv
4. `create_endpoint()` # state mutating but the state is checkpointed to internal kv.

After restart, we don't want to re-run task 3 and 4 because the state is already reconstructed. 

## Serve Replica (e.g. running ML models)
1. `__init__`
2. `handle_request` # takes longer to run
3. `handle_request` # finished
4. ...

After restart, we don't want to rerun the computation for task 3 but we do want to retry task 2. Additionally, re-running task 3 will slow down the replica and do unnecessary compute work. 

## Sidenote
I understand the library can adopt to any ray core semantic changes. but it does seem like completed actor task won't be executed is a good semantic guarantee.

Reply:
I see. I think there are still use cases for re-executing the out-of-order tasks, but I guess we can try this first.

Actually, I don't really understand the serve controller use case. Why would 3 and 4 get re-executed?

Can you also update the docs here to explain what the guarantees are for async actors? https://docs.ray.io/en/master/fault-tolerance.html#actors

Reply:
Not sure the controller case is relevant. Those methods are effectively synchronous anyways because they hold a lock during execution so they won't happen out of order.